### PR TITLE
add maker for rust and fix gcc maker exe

### PR DESCRIPTION
**Pull request by natemara**
The `gcc` maker had its exe field set to `clang`. This PR fixes that.

I also added a basic maker for the Rust programming language.

### CODE REVIEW

File: autoload/neomake/makers/c.vim [hunk 1]
```diff
@@ -25,7 +25,7 @@ endfunction
 
 function! neomake#makers#c#gcc()
     return {
-        \ 'exe': 'clang',
+        \ 'exe': 'gcc',
```
Comment (benekastah) [inline autoload/neomake/makers/c.vim]: Actually the exe field can be omitted when it matches the maker name.

**Comment by natemara** [comment]
There you go

File: autoload/neomake/makers/rust.vim [hunk 2]
```diff
@@ -0,0 +1,19 @@
+" vim: ts=4 sw=4 et
+
+function! neomake#makers#rust#EnabledMakers()
+    return ['rustc']
+endfunction
+
+function! neomake#makers#rust#rustc()
+    return {
+        \ 'args': ['-o /dev/null'],
```
Comment (benekastah) [inline autoload/neomake/makers/rust.vim]: Each arg should take its own entry in the list. Also, /dev/null is env specific, so we have a utility for that. Here is a corrected version:

```
\ 'args': ['-o', neomake#utils#DevNull()],
```

**Comment by natemara** [comment]
Fixed the `rustc` maker, added a `cargo` maker. Had a question about this though. `cargo` is a project/package manager, and is to `rustc

**Comment by benekastah** [comment]
Good question. There isn't currently a good place to put project-level makers. Issue #47 was opened to address this. Your cargo maker will actually work if you do this:

```
let g:neomake_cargo_maker = neomake#makers#rust#cargo()
```

Then `:Neomake! cargo` will run `cargo build` and put the results in the quickfix list. This isn't quite as good as the file-level makers yet. See #46 for that.

For now, perhaps you could add a TODO note referencing #47 by the cargo maker and leave it out of the enabled makers list.

**Comment by natemara** [comment]
how's this?

**Comment by natemara** [comment]
This actually doesn't work. When I run `cargo build`, it only runs `cargo`.

**Comment by natemara** [comment]
Okay, I got it working by putting `let g:neomake_cargo_maker = neomake#makers#rust#cargo()` in my nvimrc, but the cargo maker doesn't populate the location list. Is there a bigger difference between `Neomake` and `Neomake!`?

**Comment by benekastah** [comment]
I should have been more clear. The `g:neomake_cargo_maker` shouldn't be defined in any file in the autoload directory, since it may not be available (depending on whether that code has loaded). I was imagining individual users would configure that themselves until we make project-level makers easier to share.

`:Neomake!` is different from `:Neomake` in that it doesn't pass the current file as the last argument. Additionally, it puts its results in the quickfix list, not the location list. The reason for this is that the location list is associated with a window (a better fit when getting info on an individual file), whereas the quickfix list is global (better when running something at the project level). `:cwindow` should show you what you're looking for.

**Comment by natemara** [comment]
`cwindow` does show the compiler output, but the lines with errors aren't highlighed. Is this to be expected?

**Comment by benekastah** [comment]
Yup, #46 is tracking that issue. Once you remove the g:neomake_cargo_maker bit from this PR, I'll happily merge. Thanks for doing this!

**Comment by natemara** [comment]
done.

**Comment by benekastah** [comment]
#47 is fixed now. `let g:neomake_cargo_maker = neomake#makers#rust#cargo()` will no longer work. Instead, you can just do `:Neomake! cargo`. You can also do:

```
let g:neomake_enabled_makers = ['cargo']
```

Then, just running `:Neomake!` will run cargo for you.

**Pull request merged**